### PR TITLE
Fix ExponentialDecayLengthPenalty negative logits issue

### LINKED ISSUE #25416: [BUG] `ExponentialDecayLengthPenalty` decreases negative scores
### System Info

- `transformers` version: 4.31.0
- Platform: Linux-5.15.109+-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.0.1+cu118 (False)
- Tensorflow version (GPU?): 2.12.0 (False)
- Flax version (CPU?/GPU?/TPU?): 0.7.1 (cpu)
- Jax version: 0.4.14
- JaxLib version: 0.4.14
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

https://colab.research.google.com/drive/1_Ur0sQhrUan68OXDlFK0SGQfDhwlb8Pc?usp=sharing

#### Issue Explanation
`ExponentialDecayLengthPenalty` is intended to exponentially increase the score of the eos_token_id after start_index has been reached, allowing generating shorter sequences without having a hard cutoff. (Original PR: https://github.com/huggingface/transformers/pull/15245 )

When working with shorter sequences it doesn't necessarily cut the sequence, no matter how large the decay factor is.
In the line below, the processor attempts to increase the score of EOS. However when EOS score is negative, this actually decreases the score, as the exponent will be positive.
As I understand, giving a negative decay factor won't work as well due to the power. Due to this it will only succeed if EOS becomes positive.
https://github.com/huggingface/transformers/blob/f1732e1374a082bf8e43bd0e4aa8a2da21a32a21/src/transformers/generation/logits_process.py#L982

#### Proposed solution
In the attached [Colab notebook](https://colab.research.google.com/drive/1_Ur0sQhrUan68OXDlFK0SGQfDhwlb8Pc?usp=sharing), I added a proposed solution. If the score is negative, we can compute the expected penalty added to EOS if the score was positive, and add that to the original negative score.

If this is acceptable, I'd like to create a PR to fix the issue!

### PR DESCRIPTION
**Pull request by pokjay**
# What does this PR do?

In cases where the model logits are negative, ExponentialDecayLengthPenalty decreases the score of eos_token_id instead of increasing it.
To fix this issue we compute the penalty of the absolute value and add it to the original score, as described in #25416
The test was updated to check for negative logits.

In addition, this PR updates the class documentation and adds examples, as part of #24783

Fixes #25416

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@gante

### CODE REVIEW

**Review by gante — APPROVED** [review]
Thank you for detecting the issue, proposing a solution, and implementing it @pokjay 🔥 Contributions like yours make the ecosystem thrive!

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -1230,8 +1230,9 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class ExponentialDecayLengthPenalty(LogitsProcessor):
     r"""
-    [`LogitsProcessor`] that exponentially increases the score of the eos_token_id after regulation_start has been
-    reached.
+    [`LogitsProcessor`] that exponentially increases the score of the eos_token_id after start_index has been reached.
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:1233]:
```suggestion
    [`LogitsProcessor`] that exponentially increases the score of the `eos_token_id` after `start_index` has been reached.
```

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -1230,8 +1230,9 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+    This allows generating shorter sequences without having a hard cutoff, allowing the eos_token to be predicted in a
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:1234]:
```suggestion
    This allows generating shorter sequences without having a hard cutoff, allowing the `eos_token` to be predicted in a
```

**Review by ArthurZucker — APPROVED** [review]
Looks good to me! Thanks for the contribution 🤗

**Comment by pokjay** [comment]
Thanks @gante and @ArthurZucker !

Could you please approve the documentation workflow? I’d like to verify the documentation looks as expected

**Comment by pokjay** [comment]
@gante Do I need to do anything more to get this merged?

**Comment by ArthurZucker** [comment]
You could take into account the changes I suggested 😉 code related variable name are always put aroung codeblocks in our doc!

**Comment by pokjay** [comment]
@ArthurZucker oops, totally missed those! Committed those fixes now, Thanks!

**Review by ArthurZucker — APPROVED** [review]
CI should go green once you rebase to main!

**Comment by gante** [comment]
@pokjay since we last spoke, we've added this file to the list of files to be doctested in our PR CI -- it seems like your example's outputs don't match the hardcoded outputs. 

Would you be able to double-check that? :)

**Comment by pokjay** [comment]
@gante Fixed the issues, all checks pass!

**Comment by gante** [comment]
@pokjay thank you for iterating 💛
